tools/__tool

```diff
@@ -0,0 +1,311 @@
+        printf '""'
+        return
+    elif test "$__quote_tmp" = "$__quote_input"; then
+        # No special chars, so don't quote it:
+        printf %s "$__quote_input"
+        return
+    else
+        # Special chars. Escape any inner quotes, then wrap in quotes
+        __strsub_inplace __quote_input "'" "'\\''"
+        printf "'%s'" "$__quote_input"
+    fi
+}
+
+__alias() {
+    command='' name="$1"
+    shift
+    for x in "$@"; do
+        x=$(__quote "$x") || return
+        command="$command$x "
+    done
+    command="$command\"\$@\""
+    filename="/usr/local/bin/$name"
+
+    echo "#!/usr/bin/env sh" > "$filename" || return
+    echo "$command" >> "$filename" || return
+    chmod a+x "$filename" || return
+    echo "Created command alias '$filename' of [$command]" 1>&2 || return
+}
+
+__bool() {
+    bool=$(echo "$1" | __str upper)
+    case "$bool" in
+        0|OFF|NO|FALSE|N|IGNORE|NOTFOUND|""|*-NOTFOUND)
+            return 1;;
+        *)
+            return 0;;
+    esac
+}
+
+__boolstr() {
+    if __bool "$@"; then
+        printf "true"
+    else
+        printf "false"
+    fi
+}
+
+__have_command() {
+    __silently type "$1"
+}
+
+__do() {
+    printf '%s\n' "$1"
+    shift
+    local output
+    output="$("$@")"
+    rc=$?
+    if [ $rc != 0 ]; then
+        printf %s "$output"
+        return $rc
+    fi
+}
+
+# Install the named packages using the platform's package manager.
+# This script "just does the right thing" to install a package as a step in a
+#   caching container build, as a bare "pkg-tool install" might be insufficient
+#   or subtly broken, and we only really care about installing a list of packages
+#
+# Usage: <script> [pkgs...]
+__install() {
+    if test -f /etc/debian_version ; then
+        # We *must* do a repo update as part of the install step in a single cache
+        # layer, as a cached apt-get update can become stale while a later
+        # apt-get install is invalidated
+        apt-get -y update || return
+        # DEBIAN_FRONTEND suppresses all interactivity. For some reason some
+        # packages like to prompt as part of their config, even if we're not on
+        # a TTY.
+        test $# = 0 || env DEBIAN_FRONTEND=noninteractive \
+            apt-get -y install -- "$@" || return
+    elif test -f /etc/redhat-release || grep 'ID="amzn"' /etc/os-release >/dev/null 1>&2; then
+        if test -f /usr/bin/dnf; then
+            # 'dnf' will "do the right thing"
+            test $# = 0 || dnf install -y "$@" || return
+        elif test -f /usr/bin/yum; then
+            test $# = 0 || yum install -y -- "$@" || return
+            # 'yum' happily ignores missing packages. Use 'rpm -q' to check that
+            # everything we requested actually got installed.
+            test $# = 0 || if ! rpm -q -- "$@"; then
+                echo "$0: Failing because one or more packages requested are not available"
+                return 1
+            fi
```

Comment:
```suggestion
            test $# = 0 || yum install -y -- "$@" || return
```

`yum install` ignoring missing packages is supposedly fixed for [RHEL 6 and newer](https://bugzilla.redhat.com/show_bug.cgi?id=736694#c4):

> Yeh, we did fix this for RHEL-6 ... so now both "yum list" and "yum install" will exit 1 if all their arguments don't exist.

I do not think any of our RHEL-like OS images are old enough for this yum behavior to manifest, e.g. with AlmaLinux 8 (oldest RHEL-like in our list):

```
$ podman run --rm almalinux:8 bash -c 'yum install -q -y nonexist && echo Success || echo Failure'
Error: Unable to find a match: nonexist
Failure
```